--- Makefile	2010-09-11 00:46:02.000000000 +0200
+++ Makefile.patched	2015-01-01 00:00:00.000000000 +0100
@@ -68,8 +68,9 @@
 	cmp sample2.tst sample2.ref
 	cmp sample3.tst sample3.ref
 	@cat words3
+	@rm -f *.o
 
-install: bzip2 bzip2recover
+install:
 	if ( test ! -d $(PREFIX)/bin ) ; then mkdir -p $(PREFIX)/bin ; fi
 	if ( test ! -d $(PREFIX)/lib ) ; then mkdir -p $(PREFIX)/lib ; fi
 	if ( test ! -d $(PREFIX)/man ) ; then mkdir -p $(PREFIX)/man ; fi
@@ -89,6 +90,7 @@
 	chmod a+r $(PREFIX)/include/bzlib.h
 	cp -f libbz2.a $(PREFIX)/lib
 	chmod a+r $(PREFIX)/lib/libbz2.a
+	if ( test -e libbz2.so ) ; then cp -af libbz2.so* $(PREFIX)/lib ; fi
 	cp -f bzgrep $(PREFIX)/bin/bzgrep
 	ln -s -f $(PREFIX)/bin/bzgrep $(PREFIX)/bin/bzegrep
 	ln -s -f $(PREFIX)/bin/bzgrep $(PREFIX)/bin/bzfgrep
--- Makefile-libbz2_so	2010-09-11 01:07:52.000000000 +0200
+++ Makefile-libbz2_so.patched	2015-01-01 00:00:00.000000000 +0100
@@ -37,8 +37,13 @@
 all: $(OBJS)
 	$(CC) -shared -Wl,-soname -Wl,libbz2.so.1.0 -o libbz2.so.1.0.6 $(OBJS)
 	$(CC) $(CFLAGS) -o bzip2-shared bzip2.c libbz2.so.1.0.6
+	rm -f libbz2.so
+	rm -f libbz2.so.1
 	rm -f libbz2.so.1.0
 	ln -s libbz2.so.1.0.6 libbz2.so.1.0
+	ln -s libbz2.so.1.0 libbz2.so.1
+	ln -s libbz2.so.1 libbz2.so
+	rm -f $(OBJS)
 
 clean: 
 	rm -f $(OBJS) bzip2.o libbz2.so.1.0.6 libbz2.so.1.0 bzip2-shared
